packages/mobx-state-tree/src/core/action.ts

```diff
@@ -126,41 +136,94 @@ export function decorate<T extends Function>(middleware: IMiddlewareHandler, fn:
+    const middleware: IMiddleware = { handler, includeHooks: true }
     if (fn.$mst_middleware) fn.$mst_middleware.push(middleware)
     else fn.$mst_middleware = [middleware]
     return fn
 }
 
-function collectMiddlewareHandlers(
+function collectMiddlewares(
     node: ObjectNode,
     baseCall: IMiddlewareEvent,
     fn: Function
-): IMiddlewareHandler[] {
-    let handlers: IMiddlewareHandler[] = (fn as any).$mst_middleware || EMPTY_ARRAY
+): IMiddleware[] {
+    let middlewares: IMiddleware[] = (fn as any).$mst_middleware || EMPTY_ARRAY
     let n: ObjectNode | null = node
     // Find all middlewares. Optimization: cache this?
     while (n) {
-        if (n.middlewares) handlers = handlers.concat(n.middlewares)
+        if (n.middlewares) middlewares = middlewares.concat(n.middlewares)
         n = n.parent
     }
-    return handlers
+    return middlewares
 }
 
 function runMiddleWares(node: ObjectNode, baseCall: IMiddlewareEvent, originalFn: Function): any {
-    const handlers = collectMiddlewareHandlers(node, baseCall, originalFn)
+    const middlewares = collectMiddlewares(node, baseCall, originalFn)
     // Short circuit
-    if (!handlers.length) return mobxAction(originalFn).apply(null, baseCall.args)
+    if (!middlewares.length) return mobxAction(originalFn).apply(null, baseCall.args)
     let index = 0
+    let result: any = null
 
     function runNextMiddleware(call: IMiddlewareEvent): any {
-        const handler = handlers[index++]
-        if (handler) return handler(call, runNextMiddleware)
-        else return mobxAction(originalFn).apply(null, baseCall.args)
+        const middleware = middlewares[index++]
+        const handler = middleware && middleware.handler
+        let nextInvoked = false
+        let abortInvoked = false
+
+        function next(call: IMiddlewareEvent): void
+        function next(call: IMiddlewareEvent, callback: (value: any) => any): void
+        function next(call: IMiddlewareEvent, callback?: (value: any) => any) {
+            nextInvoked = true
+            // the result can contain
+            // - the non manipulated return value from an action
+            // - the non manipulated abort value
+            // - one of the above but manipulated through the callback function
+            if (callback) {
+                result = callback(runNextMiddleware(call) || result)
+            } else {
+                result = runNextMiddleware(call)
+            }
+        }
+        function abort(value: any) {
+            abortInvoked = true
+            // overwrite the result
+            // can be manipulated through middlewares earlier in the queue using the callback fn
+            result = value
+        }
+        const invokeHandler = () => {
+            handler(call, next, abort)
+            if (process.env.NODE_ENV !== "production") {
+                if (!nextInvoked && !abortInvoked) {
+                    const node = getStateTreeNode(call.tree)
+                    fail(
```

Comment:
I think this is a breaking change now? 
What we could do is, `middleware.length === 2 ? warn(message) /* a legacy decorator */ : fail(message)`?